### PR TITLE
Dev summarise

### LINKED ISSUE #6: Add save/load for graph tool

### LINKED ISSUE #7: Add save/load for networkx

### LINKED ISSUE #10: Add summary for networkx

### PR DESCRIPTION
**Pull request by bruhad-dave**
Closes #6 , closes #7 , closes #10

### CODE REVIEW

File: pp_netlib/network.py [hunk 1]
```diff
@@ -175,91 +184,92 @@ def prune(self):
...
-            genome_network (graph)
-                The loaded network
-    """
+        """Load a premade graph from a network file. 
+
+        Args:
+            network_file (str/path): The file in which the prebuilt graph is stored. Must be a .gt file, .graphml file or .xml file.
+        """
         # Load the network from the specified file
+        file_name, file_extension = os.path.splitext(network_file)
```
Comment (johnlees) [inline pp_netlib/network.py:227]: First test that self.graph = None, error if not

File: pp_netlib/network.py [hunk 2]
```diff
@@ -175,91 +184,92 @@ def prune(self):
...
+                loaded_graph = self.nx.read_graphml(network_file)
+                num_nodes = len(loaded_graph.nodes())
+                num_edges = len(loaded_graph.edges())
+
+        if file_extension == ".gt":
+            loaded_graph = self.gt.load_graph(network_file)
+            num_nodes = len(list(loaded_graph.vertices()))
+            num_edges = len(list(loaded_graph.edges()))
+            if self.backend == "NX":
+                sys.stderr.write("Network file provided is in .gt format and will be loaded with graph_tool. Please convert it to a networkx graph if you'd like to manipulate or modify it.")
```
Comment (johnlees) [inline pp_netlib/network.py:243]: Need to actually add the exit here

File: pp_netlib/network.py [hunk 3]
```diff
@@ -175,91 +184,92 @@ def prune(self):
...
-            else:
-                genome_network.from_cudf_edgelist(graph_df, renumber = False)
-            sys.stderr.write("Network loaded: " + str(genome_network.number_of_vertices()) + " samples\n")
-
-        return genome_network
-
-    def add_to_network(self, datapoint):
+        if file_extension in [".csv", ".tsv", ".txt"]:
+            sys.stderr.write("The network file appears to be in tabular format, please load it as a dataframe and use the construct method to build a graph.\n")
+            sys.exit(1)
```
Comment (johnlees) [inline pp_netlib/network.py:250]: Remove for now, add in when doing cugraph

File: pp_netlib/network.py [hunk 4]
```diff
@@ -175,91 +184,92 @@ def prune(self):
...
+                num_edges = len(loaded_graph.edges())
+
+        if file_extension == ".gt":
+            loaded_graph = self.gt.load_graph(network_file)
+            num_nodes = len(list(loaded_graph.vertices()))
+            num_edges = len(list(loaded_graph.edges()))
+            if self.backend == "NX":
+                sys.stderr.write("Network file provided is in .gt format and will be loaded with graph_tool. Please convert it to a networkx graph if you'd like to manipulate or modify it.")
+        
+        self.loaded_graph = loaded_graph
```
Comment (johnlees) [inline pp_netlib/network.py:245]: should be self.graph

File: pp_netlib/network.py [hunk 5]
```diff
@@ -175,91 +184,92 @@ def prune(self):
...
+        """Load a premade graph from a network file. 
+
+        Args:
+            network_file (str/path): The file in which the prebuilt graph is stored. Must be a .gt file, .graphml file or .xml file.
+        """
         # Load the network from the specified file
+        file_name, file_extension = os.path.splitext(network_file)
+        if file_extension in [".graphml", ".xml"]:
+            if self.backend == "GT":
+                loaded_graph = self.gt.load_graph(network_file)
```
Comment (johnlees) [inline pp_netlib/network.py:230]: should all be `self.graph`

File: pp_netlib/network.py [hunk 6]
```diff
@@ -175,91 +184,92 @@ def prune(self):
...
+                num_nodes = len(list(loaded_graph.vertices()))
+                num_edges = len(list(loaded_graph.edges()))
+            if self.backend == "NX":
+                loaded_graph = self.nx.read_graphml(network_file)
+                num_nodes = len(loaded_graph.nodes())
+                num_edges = len(loaded_graph.edges())
+
+        if file_extension == ".gt":
```
Comment (johnlees) [inline pp_netlib/network.py:238]: Put this if clause first, make the others `elif`, add and `else` clause if file extension unknown

File: pp_netlib/network.py [hunk 7]
```diff
@@ -271,38 +281,49 @@ def _convert(self, intial_format, target_format):
...
-            self.graph.to_pandas_edgelist().to_csv(file_name + ".csv.gz", compression="gzip", index = False)
+            Example:
+            ```
+            graph.save("sample_graph", ".graphml")
+            ```
+
+        Raises:
+            NotImplementedError: If graph_tool is selected a backend,
+        """
+        outdir = self.outdir
```
Comment (johnlees) [inline pp_netlib/network.py:306]: Add check that self.graph is not None

File: pp_netlib/network.py [hunk 8]
```diff
@@ -271,38 +281,49 @@ def _convert(self, intial_format, target_format):
...
+
+
+        # file_name = outdir + "/" + prefix
+        # if suffix is not None:
+        #     file_name = file_name + suffix
+        # if not self.use_gpu:
+        #     if use_graphml:
+        #         self.graph.save(file_name + ".graphml", fmt = "graphml")
+        #     else:
+        #         self.graph.save(file_name + ".gt", fmt = "gt")
```
Comment (johnlees) [inline pp_netlib/network.py:327]: get rid of commented code

File: tests/test_backend_selector.py [hunk 9]
```diff
@@ -1,4 +1,5 @@
 import os
+from re import A
```
Comment (johnlees) [inline tests/test_backend_selector.py:2]: remove

**Review by johnlees — CHANGES_REQUESTED** [review]

File: README.md [hunk 10]
```diff
@@ -1,4 +1,106 @@
...
+from pp_netlib.network import Network
+
+samples_list = ["sample1", "sample2", "sample3"]
+
+#initialise a graph object, supplying three labels, outdir as Desktop,
+#graphing module as graph-tool, and not using gpu and related modules
+example_graph = Network(ref_list = samples_list, outdir = "/Users/user/Desktop", backend = "GT", use_gpu = False)
+```
+
+Graphing backend can alternatively be set as follows:
```
[the remark this answers is not in the document]
Comment (bruhad-dave) [reply README.md:14]: Done

File: README.md [hunk 11]
```diff
@@ -1,4 +1,106 @@
...
+## .load_network
+Called on empty initialized Network object. Loads a premade graph from a network file. Network file must be .gt, .graphml, or .xml if using graph_tools as the backend library; it must be .graphml, or .xml if using networkx.
+
+```
+#initialize empty Network object
+example_loaded_network = Network([])
+example_loaded_network.load_network("path/to/network_file.graphml")
+
+# access loaded data using
+example_loaded_network.loaded_graph
```
[the remark this answers is not in the document]
Comment (bruhad-dave) [reply README.md:94]: corrected

File: pp_netlib/functions.py [hunk 12]
```diff
@@ -116,4 +127,80 @@ def construct_with_networkx(network_data, vertex_labels, weights = None):
...
+
+    elif backend == "NX":
+        import networkx as nx
+        components = nx.number_connected_components(graph)
+        density = nx.density(graph)
+        transitivity = nx.transitivity(graph)
+
+        betweenness = []
+        sizes = []
+        subgraphs = [graph.subgraph(c).copy() for c in nx.connected_components(graph)]
```
[the remark this answers is not in the document]
Comment (bruhad-dave) [reply pp_netlib/functions.py:179]: done

File: pp_netlib/functions.py [hunk 13]
```diff
@@ -116,4 +127,80 @@ def construct_with_networkx(network_data, vertex_labels, weights = None):
...
+        if len(betweenness) > 1:
+            mean_bt = np.mean(betweenness)
+            weighted_mean_bt = np.average(betweenness, weights=sizes)
+        elif len(betweenness) == 1:
+            mean_bt = betweenness[0]
+            weighted_mean_bt = betweenness[0]
+
+    metrics = [components, density, transitivity, mean_bt, weighted_mean_bt]
+    base_score = transitivity * (1 - density)
+    scores = [base_score, base_score * (1 - metrics[3]), base_score * (1 - metrics[4])]
```
[the remark this answers is not in the document]
Comment (bruhad-dave) [reply pp_netlib/functions.py:193]: done

File: pp_netlib/network.py [hunk 14]
```diff
@@ -175,91 +184,92 @@ def prune(self):
...
-        # #################################
-        # #   extra little thing to       #
-        # #  write summary to plain text  #
-        # #################################
-
-        # with open(summary_file_name, "w") as summary:
-        #     summary.write(summary_contents)
-        # summary.close()
-        return
+            raise RuntimeError("Graph not found.")
```
[the remark this answers is not in the document]
Comment (bruhad-dave) [reply pp_netlib/network.py:199]: done

File: pp_netlib/network.py (hunk 1 above)
Comment (bruhad-dave) [reply pp_netlib/network.py:227]: done

File: pp_netlib/network.py (hunk 5 above)
Comment (bruhad-dave) [reply pp_netlib/network.py:230]: done

File: pp_netlib/network.py (hunk 4 above)
Comment (bruhad-dave) [reply pp_netlib/network.py:245]: done

File: pp_netlib/network.py (hunk 6 above)
Comment (bruhad-dave) [reply pp_netlib/network.py:238]: done

**Review by johnlees — APPROVED** [review]

**Pull request merged**